### ISSUE TITLE
ExpressVPN support request

### ISSUE THREAD

**Issue by ghost** [haugene/docker-transmission-openvpn#610]
I am currently using ExpressVPN and I was looking for a torrent server solution on raspberry pi.

I am new to docker with minimal programming knowledge but I’m willing to learn. :)

Could it be possible to add ExpressVPN support? I can provide my configuration profile.

**Comment by haugene** [comment]
This should work fine with ExpressVPN and Raspberry. As for the ExpressVPN support, you're not the first: https://github.com/haugene/docker-transmission-openvpn/issues?q=is%3Aissue+expressvpn

Read through some of those and update here if you have any other questions after that.

**Comment by dannotdan** [comment]
I switched from PIA to ExpressVPN a few weeks ago after moving to a country not supported by PIA. After using this container for over a year with no problems, I've spent the last couple weeks feeling like an idiot and I'm ready to move on. Every time I fix something, something else breaks. I've never successfully signed into ExpressVPN using docker-transmission-openvpn. Seriously, I know it's been dismissed before but a lot of us really need an ELI5 to make this provider work here. I'm grateful for the hard, time-consuming work that's been done but this solution isn't working.

**Comment by haugene** [comment]
Alright. Not easy figuring it out without being able to test it myself.
Anyone willing to lend me their credentials for a short while? Then I can try to make it work and make a writeup of it in the README.

**Comment by jaeeonr** [comment]
I can provide you one if this isn't enough.
```
dev tun
fast-io
persist-key
persist-tun
nobind
remote usa-newyork-ca-version-2.expressnetw.com 1195

remote-random
pull
comp-lzo no
tls-client
verify-x509-name Server name-prefix
ns-cert-type server
key-direction 1
route-method exe
route-delay 2
tun-mtu 1500
fragment 1300
mssfix 1450
verb 3
cipher AES-256-CBC
keysize 256
auth SHA512
sndbuf 524288
rcvbuf 524288
auth-user-pass

<cert>
```

**Comment by haugene** [comment]
ExpressVPN has been a topic several times, and as far as I've understood it there is a private certificate that serves as authentication? If possible I would like to borrow your account so I can log in, download a ovpn file and try to make it work. To see all that is ExpressVPN so I don't need to guess for the answers anymore ;) And then you change your password after of course.

If that's OK, send me a PM in the gitter chat linked from the README.

**Comment by dannotdan** [comment]
I'll send something shortly. Thank you!

**Comment by jaeeonr** [comment]
@haugene did you get the credentials from dannotdan? Let me know if you still need mine. :)

**Comment by haugene** [comment]
I got it from @dannotdan. Will be away until Sunday and have a look at this next week. Will update then if I need anything else.

**Comment by dannotdan** [comment]
I actually got this working! In the process, I discovered a few things:

1. Actual credentials must be provided for OPENVPN_USERNAME and OPENVPN_PASSWORD. Dummy values don't work.

2. The adjustConfigs.sh script doesn't do anything under Windows 10. My best guess is that's due to unmet prerequisites. It works fine under Ubuntu 18.04 and probably other Linuxes as well.

3. If you are using the VPN because your location blocks access to certain sites (rather than just for privacy concerns), manually edit the .ovpn file to replace the remote address with the actual IP (which can be found using ping). You may have to edit it again later if the IP changes. Note that this step does not replace running adjustConfigs.sh.

4. The instructions in README.md should be followed closely for all other points. The error messages in the logs don't always line up with the actual issues intuitively, so shortcuts may result in a whole lot of extra time tracking down the real mistake. For example, if the .ovpn file is not found you will see a TUN error.

Thanks to @haugene for all the good work in building this image!

**Comment by jaeeonr** [comment]
I got my hands on raspberry pi last night and it's working for me too.

1. modify ovpn:
`auth-user-pass /etc/openvpn/custom/auth.txt`
2. create auth.txt containing username and password
3.
```
docker run --cap-add=NET_ADMIN --device=/dev/net/tun -d \
-e LOCAL_NETWORK=192.168.0.0/24 \
              -v /home/pi/ejp.ovpn:/etc/openvpn/custom/default.ovpn \
              -v /mnt/data/Downloads/:/data \
     -v /home/pi/auth.txt:/etc/openvpn/custom/auth.txt \
              -v /etc/localtime:/etc/localtime:ro \
		-e ENABLE_UFW=true \
              -e OPENVPN_PROVIDER=CUSTOM \
              -e OPENVPN_USERNAME=user \
              -e OPENVPN_PASSWORD=pass \
              -e WEBPROXY_ENABLED=false \
              --log-driver json-file \
              --log-opt max-size=10m \
              -p 9091:9091 \
              haugene/transmission-openvpn:dev-armhf
```


Few things I've noticed:
1. Openvpn takes up 80% out of 400% (4-core) of the CPU power. (Raspberry pi 3B+)
2. Transmission WebUI lacks critical options compared to rTorrent and ruTorrent. I still haven't figured out how to change queue size and WebUI authentication.
3. Everything works. 

Thanks for the work! 👍

**Comment by haugene** [comment]
Good news 👍 
I'll gladly merge PR to update the README if there are some hints to add in order to make it easier for others.

@dannotdan Are you running the container with --dns flags? Since you have to enter the IP directly I mean.

@jaeeonr Transmission isn't as powerful as the others you mention. But there is a couple of other web ui's supported. You can try setting TRANSMISSION_WEB_UI=transmission-web-control for example. Combustion and kettu are other options and these three are all embedded in this image, see README for more info. And also, webUI authentication can be set using TRANSMISSION_RPC_AUTHENTICATION_REQUIRED=true, and TRANSMISSION_RPC_USERNAME plus TRANSMISSION_RPC_PASSWORD variables.

**Comment by dannotdan** [comment]
@haugene I was using --dns flags but, on further examination, adding the IP directly just happened to work by chance. The real issue was that specific ExpressVPN endpoints don't always connect reliably from here. I've gone back to using URLs, but I modified the .ovpn to try several if the first does not connect (by inserting additional 'remote' lines).

**Comment by haugene** [comment]
Closing this, seems like the issue is resloved - for now :)

**Comment by SyRaza** [comment]
Sorry for commenting on a closed thread, but I just need some clarification. @ghost mentions "create auth.txt containing username and password"

What exactly does that look like?

```
user=email@address.com
pass=mypass
```

Something like this?

**Comment by haugene** [comment]
Almost. Only username on line 1. Only passord on line 2.
